File: modules/secure-recorder/android/src/main/java/expo/modules/securerecorder/EventHandler.kt
package expo.modules.securerecorder

/**
 * Handles recording events for Android
 * 
 * Coordinates limit reached and error handling. Automatically stops recording when
 * limits are exceeded and notifies handlers. Manages error state transitions.
 * 
 * ANDROID SPECIFICITY:
 * - onLimitReached is suspend function (called directly from coroutine context)
 * - onStop is suspend function (no throws)
 * - Assumes stop() succeeds (coroutine context handles errors)
 */
class EventHandler(
  private val sessionId: String,
  private val outputFile: java.io.File,
  private val recordingTimer: RecordingTimer,
  private val onStop: suspend () -> String,
  private val onLimitReached: (suspend (StopReason, String, String) -> Unit)?
) {
  internal suspend fun onLimitReached(reason: StopReason) {
    // Auto-stop recording to finalize the file
    val filePath = onStop()
    
    // Notify handler if provided
    if (onLimitReached != null) {
      onLimitReached(reason, sessionId, filePath)
    }
  }
  
  internal fun onError(message: String) {
    recordingTimer.deactivate()
    // Error is logged, state updated
    // Actual cleanup will be handled by SecureRecorderModule
  }
}

File: modules/secure-recorder/ios/EventHandler.swift
import Foundation

/**
 * Handles recording events for iOS/iPadOS
 * 
 * Coordinates limit reached and error handling. Automatically stops recording when
 * limits are exceeded and notifies handlers. Manages error state transitions.
 * 
 * iOS/iPadOS SPECIFICITY:
 * - Wraps onLimitReached in Task { @MainActor } (audio callback runs on audio thread, needs main thread dispatch)
 * - onStop throws SecureRecorderError (needs try-catch)
 * - Handles stop() failure gracefully (still notifies handler with filePath if stop fails)
 */
class EventHandler {
  private let sessionId: String
  private let outputFile: URL
  private let recordingTimer: RecordingTimer
  private let onStop: () throws -> String
  private let onLimitReached: ((StopReason, String, String) async -> Void)?
  
  internal init(
    sessionId: String,
    outputFile: URL,
    recordingTimer: RecordingTimer,
    onStop: @escaping () throws -> String,
    onLimitReached: ((StopReason, String, String) async -> Void)?
  ) {
    self.sessionId = sessionId
    self.outputFile = outputFile
    self.recordingTimer = recordingTimer
    self.onStop = onStop
    self.onLimitReached = onLimitReached
  }
  
  internal func onLimitReached(reason: StopReason) {
    // Dispatch to main queue to safely call stop() (audio callback runs on audio thread)
    Task { @MainActor [weak self] in
      guard let self = self else { return }
      
      do {
        // Auto-stop recording to finalize the file
        let filePath = try self.onStop()
        
        // Notify handler if provided
        if let handler = self.onLimitReached {
          await handler(reason, self.sessionId, filePath)
        }
      } catch {
        // If stop() fails, still notify handler if possible
        if let handler = self.onLimitReached {
          await handler(reason, self.sessionId, self.outputFile.path)
        }
      }
    }
  }
  
  internal func onError(message: String) {
    recordingTimer.deactivate()
    // Error is logged, state updated
    // Actual cleanup will be handled by SecureRecorderModule
  }
}
